common/src/app/common/types/container.cljc

```diff
@@ -385,3 +385,48 @@
       (if (ctk/in-component-copy? parent)
         true
         (has-any-copy-parent? objects (:parent-id shape))))))
+
+(defn has-any-main?
+  "Check if the shape has any children or parent that is a main component."
+  [objects shape]
+  (let [instance-root (get-instance-root objects shape)]
+    (if (or (ctk/main-instance? shape) (ctk/main-instance? instance-root))
+      true
+      (some true? (map #(has-any-main? objects (get objects %)) (:shapes shape))))))
+
+(defn valid-shape-for-component?
+  "Check if a main component can be generated from this shape in terms of component anidation:
```

Comment:
We prefer the term "nesting" over "anidation", to use a consistent vocabulary.

See https://design.penpot.dev/#/workspace/9143035f-1952-806e-8002-3ff852ca3cd5/9143035f-1952-806e-8002-3ff859163e5d?page-id=1317f2cb-e04a-806a-8002-45253d6cb753

Reply:
thanks!